components/webext-storage/src/sync/outgoing.rs

```diff
@@ -0,0 +1,185 @@
+            ext_id,
+            guid,
+            data,
+            deleted,
+            last_modified: ServerTimestamp(0),
+        },
+        change_counter,
+    ))
+}
+
+/// Gets info about what should be uploaded and also records metadata about these
+/// items in a temp table. Returns a vec of the payloads which
+/// should be uploaded. record_uploaded() can be called after the upload is
+/// complete and the data in the temp table will be used to update the local
+/// store.
+pub fn get_and_record_outgoing(
+    tx: &Transaction<'_>,
+    _signal: &dyn Interruptee,
+) -> Result<Vec<ServerPayload>> {
+    // The item may not yet have a GUID (ie, it might not already be in either
+    // the mirror nor the incoming staging table.) If not for this fact, we
+    // could probably have arranged for the populating of the outgoing staging
+    // table to be done via a single `INSERT INTO ... SELECT` statement - but
+    // due to the GUID complication, and the fact we need to extract each
+    // record from this query anyway means we just do a single query, then
+    // perform the inserts one row at a time.
+    let sql = "SELECT l.ext_id, l.data, l.sync_change_counter,
+               m.guid as mirror_guid, s.guid as staging_guid
+               FROM storage_sync_data l
+               -- left joins as one or both may not exist.
+               LEFT JOIN storage_sync_mirror m ON m.ext_id = l.ext_id
+               LEFT JOIN storage_sync_staging s ON s.ext_id = l.ext_id
+               WHERE sync_change_counter > 0";
+    let elts = tx
+        .conn()
+        .query_rows_and_then_named(sql, &[], outgoing_from_row)?;
+
+    log::debug!("get_outgoing found {} items", elts.len());
+    // Now the temp table thang...
+    let tt_sql = "INSERT INTO storage_sync_outgoing_staging
+                  (guid, ext_id, data, sync_change_counter)
+                  VALUES (:guid, :ext_id, :data, :sync_change_counter)";
+    for (payload, change_counter) in &elts {
+        log::trace!("outgoing '{:?}' with counter={}", payload, change_counter);
+        tx.execute_named_cached(
+            tt_sql,
+            rusqlite::named_params! {
+                ":guid": payload.guid,
+                ":ext_id": payload.ext_id,
+                ":data": payload.data,
+                ":sync_change_counter": change_counter
+            },
+        )?;
+    }
+    Ok(elts.into_iter().map(|e| e.0).collect())
+}
+
+/// Record the fact that items were uploaded. This updates the state of the
+/// local DB to reflect the state of the server we just updated.
+/// Note that this call is almost certainly going to be made in a *different*
+/// transaction than the transaction used in `get_and_record_outgoing()`
+pub fn record_uploaded(
+    tx: &Transaction<'_>,
+    items: &[SyncGuid],
+    signal: &dyn Interruptee,
+) -> Result<()> {
+    log::debug!(
+        "record_uploaded recording that {} items were uploaded",
+        items.len()
+    );
+
+    let sql = "
```

Comment:
I was gonna suggest doing this in a trigger, but we can do it in a follow-up, or leave it be.